src/Core/Services/ExecutionHelper.cs

```diff
@@ -0,0 +1,517 @@
+            DataSource ds = _runtimeConfigProvider.GetConfig().GetDataSourceFromDataSourceName(dataSourceName);
+            IQueryEngine queryEngine = _queryEngineFactory.GetQueryEngine(ds.DatabaseType);
+
+            if (TryGetPropertyFromParent(context, out JsonElement objectValue) &&
+                objectValue.ValueKind is not JsonValueKind.Null and not JsonValueKind.Undefined)
+            {
+                IMetadata metadata = GetMetadataObjectField(context);
+                objectValue = queryEngine.ResolveObject(objectValue, context.Selection.Field, ref metadata);
+
+                // Since the query engine could null the object out we need to check again
+                // if it's null.
+                if (objectValue.ValueKind is JsonValueKind.Null or JsonValueKind.Undefined)
+                {
+                    return null;
+                }
+
+                SetNewMetadataChildren(context, metadata);
+                return objectValue;
+            }
+
+            return null;
+        }
+
+        public object? ExecuteListField(IPureResolverContext context)
+        {
+            string dataSourceName = GraphQLUtils.GetDataSourceNameFromGraphQLContext(context, _runtimeConfigProvider.GetConfig());
+            DataSource ds = _runtimeConfigProvider.GetConfig().GetDataSourceFromDataSourceName(dataSourceName);
+            IQueryEngine queryEngine = _queryEngineFactory.GetQueryEngine(ds.DatabaseType);
+
+            if (TryGetPropertyFromParent(context, out JsonElement listValue) &&
+                listValue.ValueKind is not JsonValueKind.Null and not JsonValueKind.Undefined)
+            {
+                IMetadata metadata = GetMetadata(context);
+                object result = queryEngine.ResolveList(listValue, context.Selection.Field, ref metadata);
+                SetNewMetadataChildren(context, metadata);
+                return result;
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Set the context's result and dispose properly. If result is not null
+        /// clone root and dispose, otherwise set to null.
+        /// </summary>
+        /// <param name="context">Context to store result.</param>
+        /// <param name="result">Result to store in context.</param>
+        private static void SetContextResult(IMiddlewareContext context, JsonDocument? result)
+        {
+            if (result is not null)
+            {
+                context.RegisterForCleanup(() => result.Dispose());
+                context.Result = result.RootElement;
+            }
+            else
+            {
+                context.Result = null;
+            }
+        }
+
+        private static bool TryGetPropertyFromParent(
+            IPureResolverContext context,
+            out JsonElement propertyValue)
+        {
+            JsonElement parent = context.Parent<JsonElement>().Clone();
```

Comment:
why do we need a Clone() here?

Reply:
Looks like there is no jsondocument here so just grabbing the element should be sufficient. Thanks for call out.